### ISSUE TITLE
What user-level API, textual or graphical, what implementation model, for controlling printing.

### ISSUE THREAD

**Issue by herbelin** [coq/coq#6455]
This is a dedicated page for centralizing data about what expected model for textual (synchronous) and/or graphical (asynchronous) interface for controlling the printing of term.

To start with, see these two comments ([what should `Printing All` do](https://github.com/coq/coq/issues/6424#issuecomment-352257992) and [centralizing the implementation of the control of printing options](https://github.com/coq/coq/issues/6424#issuecomment-352258851)) at #6424, or the questions on documenting printing options at #5630.

**Comment by JasonGross** [comment]
Was there a PR or issue about deactivating specific notations without deactivating all notations?

**Comment by herbelin** [comment]
Yes, #200, moved in #703. It only exported an ML API in the absence of a decision on a user API, which basically means that it is unavailable until someone implements a [plugin](https://github.com/coq/coq/pull/703#issuecomment-305954079)...

**Comment by herbelin** [comment]
Bouncing on @ejgallego's [proposal](https://github.com/coq/coq/issues/6424#issuecomment-352258851) and a [remark](https://github.com/coq/coq/issues/5630#issuecomment-337555524) from @bcpierce00, one first step could be that `Set Printing All` is globally activating all printing options (like `Set Warnings "all"` does ), which could then be deactivated or reactivated one by one., rather than the current semantics which override the action of other printing options. (And then we could seize here an opportunity for syntax uniformization between `"all"` and `All`.)

**Comment by ppedrot** [comment]
@herbelin What do you do for printing options that are not booleans? I thought there were a few of that kind.

**Comment by herbelin** [comment]
Do you mean the tables for what is printed in `let` or `if` style?

Somehow the status of the first one is unclear considering the presence of `let 'pat`.

But anyway, we can consider that there is boolean option which governs the use of `let` or `if` globally for all constants of the table. (Or we can also have an even more refined mechanism if one wants.)

**Comment by ejgallego** [comment]
@ppedrot a record should work in that case.

**Comment by ejgallego** [comment]
Ah, I see what you mean.

**Comment by ejgallego** [comment]
Dear @herbelin, thanks for opening this issue, and sorry for the long
delay in adding my thoughts about the possible printing-related
roadmaps.

First, let me note that we already have 2 CEPs
https://github.com/coq/ceps/pull/26
https://github.com/coq/ceps/pull/19 and an issue
https://github.com/coq/coq/pull/505 discussing some ideas wrt to
printing. This comment should supersede all my comments in such prior
entries.

I will comment on 3 topics: a) document format, or what kind of
objects should the printing system build, b) AST-synthesis, or how
does Coq recovers information lost in the elaboration phase; and c)
AST-handling, for clients willing to have a low-level view on Coq's
documents themselves.

### Document Format

I think that a key question in the printing roadmap is what is the
"Coq Document Format". As of today, Coq's printing system produce
documents of type `Pp.t`, which is more or less a reification of
OCaml's `Format`'s own type. IMHO, this is a good base, but not enough
for modern printing needs. In particular, it has some character-based
features vs a more desirable pure box-based model. This means that
taking advantage of more capable backends is hard.

I think that we should extend Coq's `Pp.t` so it is better adapted to
display in richer formats such as HTML. IMHO, the box-based model that
modern HTML uses seems hard to beat, both in support,
future-proofness, and capabilities.

Also, the HTML box-model provides a good base to render to other
formats such as TeX. Some discussion has already happened in #505, and
indeed I believe a good roadmap could be:

- move `Pp.t` to a full box-based model. This means more "glue":
  clients would indeed now directly manipulate lists as the box
  constructors would be `hbox: t list -> t` as opposed to today's
  `hbox : t -> t`. This is a breaking change but easy to adapt to.
- enrich `Pp.t` with more semantics constructors, for example,
  `<p>`. It seems to me that boxes are good as they are now, [they
  roughly correspond to `<div>`. A `<notation>` entity would be a
  great idea, as well as some more flexible combinators, for those
  wanting to print proof rules, etc... This would enable a way richer
  notation system for users wanting it [an example is IRIS].
- The new output format would also imply a moderate-to-deep cleanup of
  the `printing/` subsystem. The current system it is character-based,
  and under the new model we cannot make that assumption
  anymore. IMVVHO this would be a very positive step, code-wise.
- Add a `Pp_html` module that would perform a straightforward
  rendering of the document to HTML, with the proper CSS rules.

The main question raised in #505 was one of low-level memory use. In
the `Box of t list` vs `Box of t bin_tree` it is true that the second
can be more efficient in memory. However, this advantage quickly
vanishes away when the document is sent to the client. In this case,
most serialization formats are efficient for lists, but terrible for
trees; also, in the first case we can serialize with near 0 allocation
whereas in the second case we would likely have to allocate a whole
new document, which introduces measurable latency for the user.

### Printing and the Term Synthesis Path

Another important question, and relevant for clients willing to show
Coq messages to the user is the path back from kernel terms
(`Constr.t`) to AST (`constr_expr`). In this area, I would like to
distinguish between _term-synthesis_, which is roughly what
corresponds to `Detype` today, and printing itself, of which
"externalization" is a core part.

IMHO term-synthesis should be by default synchronized with whatever
elaboration `Pretyping` is doing, trying to have an accurate term
"round-trip".

On the other side, I think externalization becomes less important in
the presence of a more capable output backend. With a richer print
document model, implicits and notations can be represented in the
output document, and thus should be always sent by default to the
client, which can then manage them [using a simple CSS rule for
example]. Imagine the convenience of hovering with the mouse and
getting rich information on implicits and notations.

Of course, some special use cases with large programs may require the
additions of some conditional processing, but for the large majority
of users this should not be the default.

In the protocol side, this means that printing state as such is
reduced to a minimum. And in fact, the goal would be to remove it
completely.

### AST Exposure

Another important topic for IDEs and tools is the ability to identify
and receive information over particular terms in the input AST. This
requires to preserve information about input locations and their
relation to the original AST, at least in the post-parsing, but still
symbolic phase.

The current `CAst` approach is a very small step towards this goal. As
we have discussed a few times, I think that it would be great to make
`glob_term` / `constrexpr` closer, or even unify them. There are a few
issues discussing minor changes, c.f. #6018 #6019.

Note that here some additional work to consolidate work of #914 would
be needed. This PR uses a smart hack to make detyping lazy with
minimal effort, but ultimately the processing strategy should be
consistent among the `glob_constr` and `constr_expr` layers.

Another key point to consider is keeping away from the AST "opaque"
and impure features. When interacting with clients, stateful and
imperative code is really a hecatomb.

### Conclusion

Well, that is all for now, I think I am not missing anything else in
my head. Let me know what you think. Ultimately, it seems to me that
all such design proposals tend to be quite opinionated; this means
that advancing in this area is bound to be very hard.

**Comment by herbelin** [comment]
Thanks for this interesting post. I'm not expert to comment, but, with naive eyes, moving from Pp.t to a richer type looks a good move.

> Imagine the convenience of hovering with the mouse and getting rich information on implicits and notations.

I confirm that this would be a great improvement.

One question I have though is: in a context where several GUI are developed (VScode, atom, ...), how to ensure that every feature which can be factorized between the different GUIs has been factorized. In particular, what shall typically be the API for locally showing rich information? What	API between the	user and the GUI engine? What API between the GUI engine and Coq?

> IMHO term-synthesis should be by default synchronized with whatever elaboration Pretyping is doing, trying to have an accurate term "round-trip".

This is	not easy. For instance,	I did not know how to do it with coercions. And	for pattern-matching compilation, it is only approximatively reconstructing	a term.	To have	an "accurate round-trip", more information from the source should be	kept in	the elaborated term.

**Comment by ejgallego** [comment]
> One question I have though is: in a context where several GUI are developed (VScode, atom, ...), how to ensure that every feature which can be factorized between the different GUIs has been factorized. In particular, what shall typically be the API for locally showing rich information? What API between the	user and the GUI engine? What API between the GUI engine and Coq?

@herbelin I think API-wise there are many choices, for example (LSP)[https://microsoft.github.io/language-server-protocol] or an evolved version of what the STM provides. However, "printing"-wise almost all modern IDEs are based on some version of the HTML + CSS + JS, so indeed, that seems like a safe bet; or course we should provide a textual renderer for other clients.

> This is	not easy. For instance,	I did not know how to do it with coercions. And	for pattern-matching compilation, it is only approximatively reconstructing	a term.	To have	an "accurate round-trip", more information from the source should be	kept in	the elaborated term.

Indeed that's a very hard problem, so the paragraph should be interpreted IMHO more as "declaration of principles" than as a hard-binding guideline.

**Comment by JasonGross** [comment]
> With a richer print
document model, implicits and notations can be represented in the
output document, and thus should be always sent by default to the
client, which can then manage them [using a simple CSS rule for
example]. Imagine the convenience of hovering with the mouse and
getting rich information on implicits and notations.
>
> Of course, some special use cases with large programs may require the
additions of some conditional processing, but for the large majority
of users this should not be the default.

I strongly disagree with this.  If every goal step requires waiting the lag time for Coq to assemble the `Set Printing All` version of the goal (often 10s of seconds, I think) that would make the UI utterly unusable for me.  I think before any change like this happens, we should have perf testing that stresses this part of the system.

**Comment by Zimmi48** [comment]
@JasonGross 
>for the large majority of users this should not be the default

I think "the large majority of users" does not include users such as yourself.

**Comment by ejgallego** [comment]
> I strongly disagree with this.

With what concretely? What percentage of Coq users do you estimate to have the special needs you have?



> If every goal step requires waiting the lag time for Coq to assemble the `Set Printing All` version of the goal (often 10s of seconds, I think) that would make the UI utterly unusable for me.

I am not sure what are you measuring here, care to share your perf numbers?

Keep in mind that many components are involved in the printing path; I wouldn't be surprised if emacs and coqide are eating most of your CPU in this case [and yes, I've done some experiments, and yes, in Coq the biggest culprit for large goals is the terrible XML rendering which this proposal would improve by at least one order of magnitude]

Anyways your use case would still be supported, in this case such information would be just be absent from the final document.

**Comment by ejgallego** [comment]
Nobody can beat for example the Chrome layout engine at rendering large stuff, not the `Format` module for sure.

**Comment by JasonGross** [comment]
> With what concretely? 

Perhaps I was too strong in my disagreement.  If there continue to exist reasonable editors where the printing path does not get slower (even if I have to set some options to get the Coq-side conditional processing), I'm fine with that.  I'm opposed to losing the current possibilities of Coq-side decrease in the amount of text.

> I am not sure what are you measuring here, care to share your perf numbers?

The amount of time between `Show` and when text starts appearing in the `*coq*` buffer in PG (not the xml version).  Note that even after text starts appearing, I often have to wait many seconds while it scrolls through tens or hundreds of thousands of lines of text.

**Comment by ejgallego** [comment]
> I'm opposed to losing the current possibilities of Coq-side decrease in the amount of text.

Certainly at a first stage we should keep all the current options regarding term non-elaboration. In particular what's the most problematic part for you? Notations? Implicits?

> The amount of time between `Show` and when text starts appearing in the `*coq*` buffer in PG

Well indeed we should get more concrete numbers on where is time spent, but printing per-se may not be the culprit here, I wouldn't be surprised if there are some superlinear parts in `Constrextern`.

**Comment by herbelin** [comment]
> @herbelin I think API-wise there are many choices, for example LSP or an evolved version of what the STM provides.

So, you mean Coq playing the role of a "language server" which would be able to treat requests like "user asked me what notation or implicit arguments expansions are possible at position X-Y, tell me what items I should show to the users in response to the right-click", or "please return the ast where the subterm at loc X-Y has its notation expanded"?

Or you mean an intermediate "language server" on top of Coq?

> I wouldn't be surprised if there are some superlinear parts in Constrextern.

Yes, there are probably improvements to get. For instance in ruling out more quickly the notations which have no chance to apply (rather than doing a sequential pattern-matching).

**Comment by ejgallego** [comment]
> So, you mean Coq playing the role of a "language server" which would be able to treat requests like "user asked me what notation or implicit arguments expansions are possible at position X-Y, tell me what items I should show to the users in response to the right-click", or "please return the ast where the subterm at loc X-Y has its notation expanded"?
> 
> Or you mean an intermediate "language server" on top of Coq?

I don't see both options as different. As of today, `coqidetop` is a language server; it sits on top of "Coq" proper, but we consider part of Coq. After all, a language server is little more than an specialized toplevel.

> Yes, there are probably improvements to get. For instance in ruling out more quickly the notations which have no chance to apply (rather than doing a sequential pattern-matching).

I'd be cool to try to list all the potential improvements we could get here. It could be good material for a Summer of Code project.

**Comment by herbelin** [comment]
> I don't see both options as different. As of today, coqidetop is a language server; it sits on top of "Coq" proper, but we consider part of Coq. After all, a language server is little more than an specialized toplevel.

Are you saying that we could try a first attempt at such "local notation expansion" feature already with `coqidetop`?

Let's for instance assume a model where, for efficiency, `constrextern.ml` attaches in advance IDs for possible alternative renderings which the GUI could request on user demand. E.g. if the term to display is `eq nat x 1`, `constrextern.ml` would produce an ast for `x = 1`, attach to the subterm `1` an ID which, when requested, will return `S O`, and attach to the subterm `x = 1` another ID which, when requested, will return `eq x 1` together with yet another ID, which when requested in turn, would return the value of the implicit argument `nat`.

Or maybe, building at once the corresponding tree of possibilities?

Is that the kind of things that you have in mind, or differently?

**Comment by ejgallego** [comment]
> Or maybe, building at once the corresponding tree of possibilities?
> 
> Is that the kind of things that you have in mind, or differently?

Indeed; in my experience tho we should try first a model where for example notation expansion is already present in the document sent to the client. State [and for example, keeping ids] in the language server has been proven to be very difficult to handle and most modern protocols [including Isabelle and LSP] avoid it.

Thus, the client would get something such as
```xml
<notation><form>#1 = #2</form><render>x = 1</render></notation><constr>eq nat <tag id="1">x</tag> <tag id="2">1</tag></constr>
```
and that could be even mapped to html directly. You could say that `contr` nodes have `display:none` for example. Note that this is bad mockup, I have something more realistic somewhere in my HD.

> Are you saying that we could try a first attempt at such "local notation expansion" feature already with `coqidetop`?

In this particular case I think this feature should be a core part of the printer / `Pp.t` type; all server/clients would get it for free.

Regarding `coqidetop` I think the serialization method is hard to evolve work-with so I am personally using (my own server)[https://github.com/ejgallego/coq-serapi] these days. One feature of such server is that it allows to select the output format [for example it has a hack named `CoqTex`]

**Comment by herbelin** [comment]
> In this particular case I think this feature should be a core part of the printer / Pp.t type; all server/clients would get it for free.

This would be a radically new `Pp.t` then if the structure allow to superimposes different alternative printing of the same expression? But why not, this is justified by the providing the new feature.

What's the exact positioning of Serapi by the way? Is it an alternative to the `xmlprotocol.ml` decoding/encoding used by CoqIDE?

**Comment by ejgallego** [comment]
> This would be a radically new `Pp.t` then if the structure allow to superimposes different alternative printing of the same expression? But why not, this is justified by the providing the new feature.

Note that indeed you can implement this with the current `Pp.t` Use `Pp.tag` and make the clients filter the pp document. Indeed, what is not convenient of the `Pp.t` format is that is not properly box-based so it is a bit harder to map to HTML / TeX.

> What's the exact positioning of Serapi by the way? Is it an alternative to the `xmlprotocol.ml` decoding/encoding used by CoqIDE?

SerAPI arose as a request by some users and PG developers to have certain kind of serialization. I already had such support in JsCoq [in fact for display trick purposes] I spun it off thinking it would be useful.

Indeed SerAPI is an "alternative" [more of an extension] that provides a more comprehensive API and tries to help clients with some complex operations in the current API.

**Comment by ejgallego** [comment]
Encoding / decoding wise it uses meta-programming so we remove quite a large part of the programming burden these kind of programs carry with them.

**Comment by Zimmi48** [comment]
@ejgallego side-question: how hard would it be to merge SerAPI in Coq and have CoqIDE use it (and deprecate the XML protocol)?

**Comment by ejgallego** [comment]
> @ejgallego side-question: how hard would it be to merge SerAPI in Coq and have CoqIDE use it (and deprecate the XML protocol)?

I dunno, I am not sure I understand the question about merging, you would just copy the folder over? However the development policies of Coq and SerAPI are very different for this work maybe.

As for CoqIDE, there already was a prototype, with the current version it wouldn't be so hard to port as SerAPI is a superset of the XML protocol, however future versions will make the protocol quite different.

**Comment by Zimmi48** [comment]
I was thinking making SerAPI a first-class, official protocol. But yes, I understand that you prefer to work on it on the side as it gives you greater flexibility regarding the development strategies.
